Task #8353 Add PageableAdvancedDto as Response Wrapper in the ExportSettingController.

ChernenkoVitaliy opened the pull request:
# GreenCity PR
## Issue Link 📋
<a href="https://github.com/ita-social-projects/GreenCity/issues/8353">#8353</a>

## Changed

- Added _Pageble_ as method param in the method _selectFromTable_ in the `ExportSettingsController`;
- Refactored unit tests due to new method params in the `ExportSettingsControllerTest` class;
- Added method for receiving `PageableAdvancedDto` as TableRows in the `ModelUtils`;
- Refactored _selectFromTable_ method in the `ExportSettingsService` interface;
- Refactored realisation of the method _selectFromTable_ in the `ExportSettingsServiceImpl` and added private helping method _populatePageableDto_;
- Added method `countRowsInTable` in the `ExportSettingsRepo` interface and implementation in the `ExportSettingsRepoImpl`;
- Added new integration tests in the `ExportSettingsRepoImplTest` for covering the new method;
- Added constant _SELECT_COUNT_FROM_ in the `AppConstant` class;

## Summary Of Changes 🔥
Added pageable functionality for selecting data from the DB;

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added the ability to count rows in any database table for export settings.
  - Responses for table data export are now paginated and provide advanced pagination metadata.

- **Refactor**
  - Updated endpoints and service methods to use standard pagination parameters (`page` and `size`) instead of custom limit/offset.
  - Improved response structure for table data export to a more flexible and generic format.

- **Bug Fixes**
  - Enhanced validation for table names in API requests.

- **Tests**
  - Added and updated tests to cover new pagination logic and row-counting functionality.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ChernenkoVitaliy committed:
Added Pageble as method param in the method selectFromTable in the ExportSettingsController;
Refactored unit tests cause of new method params in the ExportSettingsControllerTest class;
Added method for receiving PageableAdvancedDto as TableRows in the PageableAdvancedDto;
Refactored selectFromTable method in the ExportSettingsService interface;
Refactored realisation of the method selectFromTable in the ExportSettingsServiceImpl and added provate helping method populatePageableDto;
Added method countRowsInTable in the ExportSettingsRepo interface and implementation in the ExportSettingsRepoImpl;
Added new integration tests in the ExportSettingsRepoImplTest for cover new method;
Added constant SELECT_COUNT_FROM in the AppConstant class;

ChernenkoVitaliy committed:
Returned removed constant EMPTY_STRING in the AppConstant;

maks741 commented on core/src/main/java/greencity/controller/ExportSettingsController.java:
```diff
@@ -68,20 +74,22 @@ public ResponseEntity<TablesMetadataDto> getTablesInfo() {
-    public ResponseEntity<TableRowsDto> selectFromTable(@Valid TableParamsRequestDto tableParams) {
-        return ResponseEntity.ok(exportSettingsService.selectFromTable(tableParams));
+    public ResponseEntity<PageableAdvancedDto<Map<String, String>>> selectFromTable(
+        @Pattern(regexp = "^(?!_)[a-z]+(?:_[a-z]+){0,10}(?<!_)$",
```
maybe it would be better to extract the regex to a separate file containing constants

ChernenkoVitaliy replied:
It is a good idea!
Fixed.

ChernenkoVitaliy committed:
Added missed unit tests for ExportSettingsRepoImpl and ExportSettingsServiceImplTest;
Small refactor of method countRowsInTable in the ExportSettingsRepoImpl;
Extracted regex for valid table name in the AppConstant;

ChernenkoVitaliy committed:
Fixed Sonar issues:

ChernenkoVitaliy committed:
Improved test coverage specialy for sir Sonar;

ChernenkoVitaliy committed:
Fixed bug in the populateTableRowDto method  in the ExportSettingsServiceImplTest;

ChernenkoVitaliy commented on service/src/test/java/greencity/service/ExportSettingsServiceImplTest.java:
```diff
@@ -77,4 +104,19 @@ void getEnvironmentVariablesTest() {
+            tableData.add(row);
+        }
+        return new TableRowsDto("users", tableData);
+    }
```
Thanks for pointing that out. It was a bug.
Fixed.

ChernenkoVitaliy commented on service/src/main/java/greencity/service/ExportSettingsServiceImpl.java:
```diff
@@ -23,11 +27,14 @@ public TablesMetadataDto getTablesMetadata() {
-    public TableRowsDto selectFromTable(TableParamsRequestDto tableParams) {
-        return exportSettingsRepo.selectPortionFromTable(tableParams.tableName(), tableParams.limit(),
-            tableParams.offset());
+    @Transactional(readOnly = true)
+    public PageableAdvancedDto<Map<String, String>> selectFromTable(String tableName, Pageable pageable) {
+        int totalElements = exportSettingsRepo.countRowsInTable(tableName);
+        TableRowsDto data = exportSettingsRepo.selectPortionFromTable(
+            tableName, pageable.getPageSize(), (int) pageable.getOffset());
+
+        return populatePageableDto(totalElements, pageable, data.tableData());
     }
```
All SQLExceptions are handled in the ExportSettingsRepoImpl class.
Then they will be handled by CustomExceptionHandler, so I think catching them here is unnecessary.

If you don't agree, please let me know.

ChernenkoVitaliy commented on core/src/main/java/greencity/controller/ExportSettingsController.java:
```diff
@@ -68,20 +74,22 @@ public ResponseEntity<TablesMetadataDto> getTablesInfo() {
+            message = ErrorMessage.INVALID_TABLE_NAME) String tableName,
+        @Parameter(hidden = true) Pageable pageable) {
+        return ResponseEntity.ok(exportSettingsService.selectFromTable(tableName, pageable));
     }
```
Thanks.
But we already have a default behavior for Pageable in the application.
If it is not passed, it will be set with page 0 and size 20 by default. Not necessary to write any annotation explicitly.
